cpp/src/arrow/dataset/discovery.cc

```diff
@@ -47,6 +46,59 @@ Result<std::shared_ptr<Schema>> SourceFactory::Inspect() {
   return UnifySchemas(schemas);
 }
 
+DatasetFactory::DatasetFactory(std::vector<std::shared_ptr<SourceFactory>> sources)
+    : sources_(std::move(sources)) {}
+
+Result<std::shared_ptr<DatasetFactory>> DatasetFactory::Make(
+    std::vector<std::shared_ptr<SourceFactory>> sources) {
```

Comment:
Is it confusing that the variable name is `sources` but they're actually `SourceFactories`?

Reply:
Agreed:
```suggestion
    std::vector<std::shared_ptr<SourceFactory>> factories) {
```